### DISCUSSION TITLE
Controller v.s. negotiator

### DISCUSSION THREAD

**Discussion by rezunli96** [yasserfarouk/scml#20] [Q&A] [answered]
Just want to confirm: for a negotiator it deals with one negotiation one at a time, but for a controller it is handling multiple offers at the same time. To be more specific, in oneshot a controller will negotiating with all agents from the other level simultaneously, but in std/col it is relatively uncertain in each call who it is dealing with right?

**Comment by yasserfarouk** [comment, marked as answer]
Your understanding of the difference between a controller and a negotiator is correct. 

A negotiator or a controller **can always** knows who (by who I men which agent) are they dealing with. The only difference is the number of negotiations it is engaged in. A negotiator enters only a single negotiation so its partner is implicit. That is why `propose()` and `respond()` methods of `SAONegotiator` has no explicit parameter specifying the partner.

* the rest of this reply is about controllers*
 
The basic controller in SCML `SAOController` always receives a `negotiator_id` which specifies the id assigned to one of its current set of negotiations. 

In the one-shot track, we guarantee that the `negotiator_id` equals the `id` of the partner in this negotiation. For example if agent x is negotiating with agent y, the negotiator_id x has for this negotiation is y and vice versa.

In std/col, this is not guaranteed but the agent can still easily get the ID of the partner it is negotiating with. In any callback for the controller, it gets a `negotiator_id` and it can use it to access the `annotations` of the current negotiation using:

```python
a = self.negotiators[negotiator_id][0].ami.annotations
```

To get the partner ID, you can use something like:

```python
partner_id = a["buyer"] if a["seller"] == self.owner.id else a["seller"]
```
The reason **we cannot match the `negotiator_id` to the partner ID in std/col** is that the agent itself creates the negotiator (either in its call to `request_negotiation(s)` or `respond_to_negotiation_request`).

**Reply by rezunli96** [reply]
I see.
Another tangential question: does one know who are default agents?

**Comment by yasserfarouk** [comment]
No you never know explicitly the type of the agent you are trading with. During development and testing, we keep part of the type name as part of the agent name for convenience and to simplify debugging. In the official competition, agent names will be random. There is no way in the system to know the type of the agent. 

Of course it is always possible to got it using supposedly private members of the world object but that is considered cheating :-)

**Reply by rezunli96** [reply]
Yes this make sense.

**Comment by rezunli96** [comment]
I have several follow-up questions: how does the order of negotiation determined? For example if I adopt negotiator instead of controller, is it that I first finish negotiation with opponent A then B? Or it could be interleaving such that I can make use of the information of negotiation with B into the one with A? And if I adopt a controller, what determines the received offers in counter_all()? For example in oneshot, if I only encounter a subset of my opponents, does it mean I will keep running negotiations with these ones until all of them finished, and then run negotiation with the rest ones?

**Reply by yasserfarouk** [reply]
> how does the order of negotiation determined?

It is random. Adopting a negotiator or a controller causes no difference here. 

The process is slightly different in oneshot and standard/collusion so I will
describe both assuming that you have suppliers [A, B, C] and consumers [X, Y,
Z].

- **Standard** You choose a set of your suppliers and/or consumers to negotiate
	with by calling `request_negotiation` or `request_negotaition**s**` (the
	earlier needs a negotiator and the later a controller/set of negotiators). No
	matter which way do you `request` negotiations, they will all run
	**concurrently** on the next simulation step. This means that they are
	interleaved. It is not possible for you to finish one negotiation before
	starting the next.
- **OneShot** You always negotiate with all of your suppliers. You do not need
to `request` negotiations. Nevertheless, the same logic apply. All negotiations
run **concurrently** (i.e. interleaved) and the order is random.

> If I adopt a controller, what determines the received offers in `counter_all()`?

Not sure that I understand the question but here is my attempt at answering it.

- `counter_all()` tries to receive **one offer from each partner**. What the
	controller tries to do is accumulating offers from partners until exactly one
	is received from each then calls `counter_all()`. The problem is that if some
	of these partners are themselves using a `SAOSyncController`, a loop may form
	(something like the philosopher's dinner problem). When the system detects
	a loop, it breaks it **by forcing `counter_all` to be called with only a
	partial set of partner offers**. Because the order of the negotiations is
	random, you cannot control which partial set will be selected in this case.

> In oneshot, if I only encounter a subset of my opponents

That never happens. In OneShot you always have one negotiation with each
partner each step. Of course you can choose to end some of them immediately.

**Reply by rezunli96** [reply]
thank you so much, this really clarifies a lot!
Few more questions about counter_all: does it synchronize the negotiation rounds of all current opponents? For example if I am A, and now X, Y has already give me the first offer. Then is counter_all suspended until Z gives me the next one? So that I always receive offers of the same round index from all X, Y, Z?

Also just to check if I understand correctly: if A uses a controller and X also uses a controller (assuming others negotiators), what could happen is that in A's counter_all it receives offers from [X, Y, Z] but for X it is [B, C] right? And I guess in that day in their counter_all they will be receiving exactly the same set of opponent ([X, Y, Z] for A and [B, C] for X) repeatedly? How are the partial sets determined?

**Comment by rezunli96** [comment]
Hi one question just pumped up. So in oneshot does negotiator_id received in propose() or respond() really equals to the id of your partner? I had a local test which shows that the negotiator_id are always something like '02Ran@0-19edad1e-db6b-4d3a-9ecb-f8423034ded7', while self.awi.my_consumers are ['SELLER'] or something like ['04Ran@1', '05Ran@1', '06Ran@1', '07Ran@1']. So what does self.awi.my_consumers or self.awi.my_supplies return and how can we get the id of my opponents?

**Reply by yasserfarouk** [reply]
They should be equal. If not, we will consider it a bug.

My guess is that this long string is the id of the opponent’s negotiator not the opponent itself.

Can you share the test?

**Reply by rezunli96** [reply]
So in my test I just each time print negotiator_id and self.awi.my_consumers and self.awi.my_suppliers each time it calls propose() and response().

**Reply by yasserfarouk** [reply]
I got it. I think I also reproduced it. Can you upgrade to the latest version on github. The last commit should solve this issue.

you may also need to upgrade negmas to the latest version on github if your other tests started to fail.

**Reply by rezunli96** [reply]
thank you!

**Reply by rezunli96** [reply]
Hi, I still get these long strings sometimes. And this time I also print self.awi.current_step, and my observation is that we get them only at day 0 and any other days seem to be fine?

**Reply by rezunli96** [reply]
hello sorry for keeping pumping up but after I fix the line that shown cause this bug, I still find them at day 0 sometimes.

**Reply by yasserfarouk** [reply]
Will add a test for that and try to reproduce the issue. By the way, can you move this into an issue instead of this discussion thread to make it visible for others who may have the same issue.

thanks
yasser

**Reply by yasserfarouk** [reply]
I moved the discussion into issue #33 .

Please use this issue for any further discussion about it
